tianshou/data/batch.py

```diff
@@ -546,14 +553,19 @@ def empty_(self) -> 'Batch':
                 continue
             if isinstance(v, Batch):
                 self.__dict__[k].empty_()
-            elif isinstance(v, np.ndarray) and v.dtype == np.object:
-                self.__dict__[k].fill(None)
             elif isinstance(v, torch.Tensor):  # cannot apply fill_ directly
-                self.__dict__[k] = torch.zeros_like(self.__dict__[k])
-            else:  # np
-                self.__dict__[k] *= 0
-                if hasattr(v, 'dtype') and v.dtype.kind in 'fc':
-                    self.__dict__[k] = np.nan_to_num(self.__dict__[k])
+                self.__dict__[k].zero_()
+            elif isinstance(v, np.ndarray):
+                if v.dtype == np.object:
+                    self.__dict__[k].fill(None)
+                else:  # np
+                    self.__dict__[k].fill(0)
+            else:
+                # scalar value
+                if isinstance(v, np.number) or isinstance(v, np.bool_):
+                    self.__dict__[k] = self.__dict__[k].__class__(0)
```

Comment:
`self.__dict__[k] *= 0` is better in that particular case, since memory allocation must be avoided.

Reply:
Scalar values in numpy are immutable. Here memory allocation cannot be avoided.
```
tmp = np.float64(1.0)
tmp
Out[160]: 1.0
tmp2 = tmp
tmp2 *= 0
tmp
Out[163]: 1.0
```

Reply:
It is an issue of numpy scalar value:
```python
In [2]: a=np.array([1,1,3])                                                                                                                                                                                         

In [3]: a[0].fill(0)                                                                                                                                                                                                

In [4]: a                                                                                                                                                                                                           
Out[4]: array([1, 1, 3])
```

Reply:
> Scalar values in numpy are immutable. 

Good to know. Indeed it almost looks like a bug for me. Anyway, I still find it clearer to do `self.__dict__[k] *= 0`, and it works with @Trinkle23897 example:
```
In [2]: a=np.array([1,1,3])

In [3]: a[0] *= 0

In [4]: a
Out[4]: array([0, 1, 3])
```

Reply:
Yes but it will fail in this case:
```python
In [82]: d=Batch(a=[1,2,3])

In [83]: d[0].empty_()
Out[83]: 
Batch(
    a: 0,
)

In [84]: d
Out[84]: 
Batch(
    a: array([1, 2, 3]),
)
```
I'm implementing empty_(index=...) and this method can avoid the above issue:
```python
d.empty_(index=0)
```

Reply:
Hum... it's nice that `d.empty_(index=0)` is working fine, but it is very unfortunate that `d[0].empty_()` is not. I think it is a major source of error for the user (as it is for numpy).

Reply:
I add a warning.